operator/internal/controller/pulumi/stack_controller.go

```diff
@@ -628,24 +629,35 @@ func (r *StackReconciler) Reconcile(ctx context.Context, request ctrl.Request) (
 			return reconcile.Result{}, err
 		}
 
-		// case stack.ProgramRef != nil:
-		// 	programRef := stack.ProgramRef
-		// 	if currentCommit, err = sess.SetupWorkdirFromYAML(ctx, *programRef); err != nil {
-		// 		r.emitEvent(instance, pulumiv1.StackInitializationFailureEvent(), "Failed to initialize stack: %v", err.Error())
-		// 		reqLogger.Error(err, "Failed to setup Pulumi workspace", "Stack.Name", stack.Stack)
-		// 		r.markStackFailed(sess, instance, err, "", "")
-		// 		if errors.Is(err, errProgramNotFound) {
-		// 			instance.Status.MarkStalledCondition(pulumiv1.StalledSourceUnavailableReason, err.Error())
-		// 			return reconcile.Result{}, nil
-		// 		}
-		// 		if isStalledError(err) {
-		// 			instance.Status.MarkStalledCondition(pulumiv1.StalledSpecInvalidReason, err.Error())
-		// 			return reconcile.Result{}, nil
-		// 		}
-		// 		instance.Status.MarkReconcilingCondition(pulumiv1.ReconcilingRetryReason, err.Error())
-		// 		// this can fail for reasons which might go away without intervention; so, retry explicitly
-		// 		return reconcile.Result{Requeue: true}, nil
-		// 	}
+	case stack.ProgramRef != nil:
+		var program unstructured.Unstructured
+		program.SetAPIVersion(v1.GroupVersion.String())
+		program.SetKind("Program")
```

Comment:
I'm curious why Unstructured and not the Program type here?

Reply:
The `SetupWorkspaceFromFluxSource` method takes in an unstructured object to obtain the `URL` and `Digest` fields. Instead of getting the object as a Program and then converting to unstructured (ie. `runtime.DefaultUnstructuredConverter.ToUnstructured(object)`), I'm just doing this directly.